src/model/idtypes.rs

```diff
@@ -0,0 +1,300 @@
+use crate::model::Type;
+use serde::{Deserialize, Serialize};
+use std::borrow::Borrow;
+use std::marker::PhantomData;
+use strum::Display;
+use thiserror::Error;
+
+// This is a sealed trait pattern implementation, it stops external code from implementing the `IdType` trait.
+// The `Sealed` trait must be in a private mod, so external code can not see and implement it.
+// See also: https://rust-lang.github.io/api-guidelines/future-proofing.html
+mod private {
+    pub trait Sealed {}
+}
+
+pub trait IdType: private::Sealed {
+    const TYPE: Type;
+}
+pub trait PlayableIdType: IdType {}
+pub trait PlayContextIdType: IdType {}
+
+impl IdType for Artist {
+    const TYPE: Type = Type::Artist;
+}
+impl PlayContextIdType for Artist {}
+impl IdType for Album {
+    const TYPE: Type = Type::Album;
+}
+impl PlayContextIdType for Album {}
+impl IdType for Track {
+    const TYPE: Type = Type::Track;
+}
+impl PlayableIdType for Track {}
+impl IdType for Playlist {
+    const TYPE: Type = Type::Playlist;
+}
+impl PlayContextIdType for Playlist {}
+impl IdType for User {
+    const TYPE: Type = Type::User;
+}
+impl IdType for Show {
+    const TYPE: Type = Type::Show;
+}
+impl PlayContextIdType for Show {}
+impl IdType for Episode {
+    const TYPE: Type = Type::Episode;
+}
+impl PlayableIdType for Episode {}
+
+pub type ArtistId = Id<Artist>;
+pub type AlbumId = Id<Album>;
+pub type TrackId = Id<Track>;
+pub type PlaylistId = Id<Playlist>;
+pub type UserId = Id<User>;
+pub type ShowId = Id<Show>;
+pub type EpisodeId = Id<Episode>;
+
+pub type ArtistIdBuf = IdBuf<Artist>;
+pub type AlbumIdBuf = IdBuf<Album>;
+pub type TrackIdBuf = IdBuf<Track>;
+pub type PlaylistIdBuf = IdBuf<Playlist>;
+pub type UserIdBuf = IdBuf<User>;
+pub type ShowIdBuf = IdBuf<Show>;
+pub type EpisodeIdBuf = IdBuf<Episode>;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Artist {}
+impl private::Sealed for Artist {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Album {}
+impl private::Sealed for Album {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Track {}
+impl private::Sealed for Track {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Playlist {}
+impl private::Sealed for Playlist {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum User {}
+impl private::Sealed for User {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Show {}
+impl private::Sealed for Show {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Episode {}
+impl private::Sealed for Episode {}
+
+/// A Spotify object id of given [type](crate::model::enums::types::Type)
+///
+/// This is a not-owning type, it stores a &str only.
+/// See [IdBuf](crate::model::idtypes::IdBuf) for owned version of the type.
+#[derive(Debug, PartialEq, Eq, Serialize)]
+pub struct Id<T> {
+    #[serde(default)]
+    _type: PhantomData<T>,
```

Comment:
Sorry for not getting your point, I am a little bit confused about the usage of `PhantomData<T>`, what kind of problem does it solve, is it fine to remove `PhantomData<T>` or replace it with another type?

Reply:
All generic type arguments must be bound to some struct field. It's a hard Rust requirement. Here type argument is just a marker, there's no useful field of this type in the structure. That's why we need zero-sized dummy types like `PhantomData<T>`. Its role is to show the compiler I really use this type argument `T`.

Reply:
You can read about `PhantomData` here: https://doc.rust-lang.org/nomicon/phantom-data.html